src/rio.c

```diff
@@ -108,19 +109,49 @@ void rioInitWithBuffer(rio *r, sds s) {
 
 /* Returns 1 or 0 for success/failure. */
 static size_t rioFileWrite(rio *r, const void *buf, size_t len) {
-    size_t retval;
+    if (!r->io.file.autosync) return fwrite(buf,len,1,r->io.file.fp);
 
-    retval = fwrite(buf,len,1,r->io.file.fp);
-    r->io.file.buffered += len;
+    size_t nwritten = 0;
+    while (len != nwritten) {
+        assert(r->io.file.autosync > r->io.file.buffered);
+        size_t nalign = (size_t)(r->io.file.autosync - r->io.file.buffered);
+        size_t towrite = nalign > len-nwritten ? len-nwritten : nalign;
+
+        if (fwrite((char*)buf+nwritten,towrite,1,r->io.file.fp) == 0) return 0;
+        nwritten += towrite;
+        r->io.file.buffered += towrite;
+
+        if (r->io.file.buffered >= r->io.file.autosync) {
+            fflush(r->io.file.fp);
+
+            size_t processed = r->processed_bytes + nwritten;
+            assert(processed % r->io.file.autosync == 0);
+            assert(r->io.file.buffered == r->io.file.autosync);
+
+#if HAVE_SYNC_FILE_RANGE
+            /* Start writeout asynchronously. */
+            if (sync_file_range(fileno(r->io.file.fp),
+                    processed - r->io.file.autosync, r->io.file.autosync,
+                    SYNC_FILE_RANGE_WRITE) == -1)
+                return 0;
 
-    if (r->io.file.autosync &&
-        r->io.file.buffered >= r->io.file.autosync)
-    {
-        fflush(r->io.file.fp);
-        if (redis_fsync(fileno(r->io.file.fp)) == -1) return 0;
-        r->io.file.buffered = 0;
+            if (processed >= (size_t)r->io.file.autosync * 2) {
+                /* To keep the promise to 'autosync', we should make sure last
+                 * asynchronous writeout persists into disk. This call may block
+                 * if last writeout is not finished since disk is slow. */
+                if (sync_file_range(fileno(r->io.file.fp),
+                        processed - r->io.file.autosync*2,
+                        r->io.file.autosync, SYNC_FILE_RANGE_WAIT_BEFORE|
```

Comment:
this seems a bit fragile and although i don't see a problem (since the code guarantees to enter this block exactly on each and every 4mb chunk), i think this can achieve the same result:
```suggestion
                        0, processed - r->io.file.autosync,
                        SYNC_FILE_RANGE_WAIT_BEFORE|
```
IMHO we can also safely drop SYNC_FILE_RANGE_WAIT_BEFORE if we want.

Reply:
since there are many asserts, i think it is right, but if you want to this change, i feel free if there is no degradation after my later tests

Reply:
i think that if you'll do that change and show that it doesn't change the performance, it'll be safer.

Reply:
@oranagra There is a small degradation. Using benchmark write some data, dumping rdb is 7.7G, there are 4 tests for dumping rdb.
|   From 0 | From (processed - autosync*2) |
|  ----  | ----  |
| 51s  | 45s |
| 48s  | 46s |
| 49s  | 46s |
| 49s  | 45s |

I have a quick look at linux code, it seems kernel iterate all pages of sync range, not sure i am right

http://lkml.iu.edu/hypermail/linux/kernel/1005.2/01845.html is the advise of Linus.

Reply:
Ok too bad. Let's stick to your version and rely on the assertions.

Reply:
@oranagra Sorry, maybe i make a mistake, my code is not clean, i changed, there is no obvious performance difference.
kernel only iterate dirty page of specific range, 
https://github.com/torvalds/linux/blob/440462198d9c45e48f2d8d9b18c5702d92282f46/mm/page-writeback.c#L2206-L2211, it fist finds dirty range and then wait pages writeback.

So you can apply your suggestions, certainly, it is safer in theory, but i can't find current PR has bugs that don't sync some pages.

Reply:
ok, even if we somehow missed something, the damage will only be performance issue due to dirty page cache..